internal/bindings/server.go

```diff
@@ -277,8 +278,17 @@ func connectWithDial(handle C.uintptr_t, address *C.char, fd *C.int) C.int {
 	defer connectLock.Unlock()
 	dial := connectRegistry[handle]
 	// TODO: make timeout customizable.
-	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
-	defer cancel()
+	ctx, cancel := context.WithTimeout(dialCtx, 5*time.Second)
+
+	// The dial function may be asynchronous so only cancel the context
+	// if we receive an error.
+	var err error
```

Comment:
Why not do this in the error handling of dial() below? Also I don't get your point about async here?

Reply:
>Why not do this in the error handling of dial() below?
No reason, just thought to put all context related things in their own block.
>Also I don't get your point about async here?
Just that `dial` isn't guaranteed to be finished before its caller `connectWithDial`.

Reply:
How can that be? And how can you test the return value in that case?

Also the scoping on err may be problematic, especially on the call to connToSocket

Reply:
So in the case of the TLS and external dial functions, (similar to LXD's `dqliteProxy`) when `dial` is run, it starts up a the `proxy` goroutine which in turn starts up two more goroutines that we expect to read and write from the connection that `dial` is returning until there's no more data.  So `dial` is returning a `net.Conn` that is still being used by a goroutine.

In practice, if we `defer cancel()` the parent context of the proxy inside of `connectWithDial`, the proxy will always end prematurely, before the `io.Copy` goroutines complete. 

>Also the scoping on err may be problematic, especially on the call to connToSocket

Not really a problem, since we would need to call a `cancel()` in the error block of `connToSocket` too.

Reply:
The err var being tested in the defer won't be the same one that is being returned from connToSocket, in that way it is a problem I think.

Reply:
OK so I've taken a closer look at this now that I'm back on my PC :)

My take from this is that in a nutshell we have two contexts at play:

1. The dial context - short lived (cancelled after 5s if not connected before)
2. The node context - long lived (cancelled when the node is stopped).

The thing I'm not quite clear on still though is that, when I call `func (s *Node) SetDialFunc(dial protocol.DialFunc, ctx context.Context)` am I providing the node context or the dial context?

Because on the one hand, its being stored in the `contextRegistry` suggesting a long-lived node context, but then `connectWithDial` is is used from the registry and named `dialCtx` (without extending it), and then has a 5s timeout added to it which is then called `ctx`. If the 5s timeout was the dialCtx, then I would have expected the naming inside `connectWithDial` to be the other way round (i.e the node ctx was loaded into `ctx` and then wrapped with a 5s timeout resulting in `dialCtx`).

I also don't understand why the `ctx, cancel := context.WithTimeout(dialCtx, 5*time.Second)` is cancelled on defer only if there has been an error. This is odd, as the context *will still* be cancelled after 5s even if there hasn't been an error (because it is `WithTimeout` type).

I think the friction here is that the formerly merged `extDialFuncWithProxy` function (https://github.com/canonical/go-dqlite/pull/179/files#diff-2b86d61eafcf5f028792eac153162dd06da4986de4d4c560d46cd51b85c75cf7R43) is returning a function that can only accept a single context (client.DialFunc) but really needs access to both; the dial context for the actual dial, and the long-lived node context for the long-lived proxy go routine that is started afterwards.

It feels to me like this could be solved by updating `makeNodeDialFunc` to accept the node's long-lived context from the `contextRegistry`, which could then be used by the `go proxy(ctx, conn, goUnix, clonedConfig)` line rather than using the dial context.

Reply:
> The thing I'm not quite clear on still though is that, when I call func (s *Node) SetDialFunc(dial protocol.DialFunc, ctx context.Context) am I providing the node context or the dial context?

Poor naming there on my part, yes the `contextRegistry` holds stores the node context which is then used as the parent context for the dial context.

> It feels to me like this could be solved by updating makeNodeDialFunc to accept the node's long-lived context from the contextRegistry, which could then be used by the go proxy(ctx, conn, goUnix, clonedConfig) line rather than using the dial context.

I like this idea, though `makeNodeDialFunc` runs before there is any notion of a node so it would need to take the app context instead.

Reply:
That would probably be fine to have an app context rather than node context, as presumably if you remove a node that will cause its sockets be closed, causing its proxy go routine to end anyway right?